### PR TITLE
Fix active VU reporting by arrival-rate executors

### PR DESCRIPTION
**Pull request by imiric**
This is a somewhat quick and dirty fix of #1977, where instead of changing the global active VUs when the VU is activated, we change it when the VU is actually running. This synchronizes it with the running count of the internal `activeVUPool`, and reports the correct value everywhere the `ExecutionState` is used (the real-time total active VU count while the test is running, in all outputs, etc.).

### CODE REVIEW

File: lib/executor/ramping_arrival_rate_test.go [hunk 1]
```diff
@@ -734,3 +734,45 @@ func TestRampingArrivalRateCornerCase(t *testing.T) {
...
+	})
+	test := setupExecutorTest(t, "", "", lib.Options{}, runner, config)
+	defer test.cancel()
+
+	getCurrActiveVUs = test.state.GetCurrentlyActiveVUsCount
+
+	engineOut := make(chan metrics.SampleContainer, 1000)
+	require.NoError(t, test.executor.Run(test.ctx, engineOut))
+
+	assert.GreaterOrEqual(t, running, int64(5))
```
Comment (codebien) [inline lib/executor/ramping_arrival_rate_test.go:777]: nit: can we check that it isn't bigger than maxVUs?

File: lib/executor/constant_arrival_rate.go [hunk 2]
```diff
@@ -264,7 +264,7 @@ func (car ConstantArrivalRate) Run(parentCtx context.Context, out chan<- metrics
 	}()
 
 	returnVU := func(u lib.InitializedVU) {
-		car.executionState.ReturnVU(u, true)
+		car.executionState.ReturnVU(u, false)
 		activeVUsWg.Done()
```
Comment (codebien) [inline lib/executor/constant_arrival_rate.go:268]: This could generate confusion in the future, we set ReturnVU with _false_ but are setting **activeVUs**Wg as Done. Should we add a small comment to remind us why are we doing it? e.g. `We set activeVUs as false because the counter is already updated by the pool`.

**Review by codebien — CHANGES_REQUESTED** [review]
>where instead of changing the global active VUs when the VU is activated, we change it when the VU is actually running

Is this logic consistent across all executors now? :thinking:

File: lib/executor/constant_arrival_rate.go (hunk 2 above)
Comment (imiric) [reply lib/executor/constant_arrival_rate.go:272]: Good point. See 8d43a834.

**Comment by imiric** [comment]
@codebien

>> where instead of changing the global active VUs when the VU is activated, we change it when the VU is actually running
>
> Is this logic consistent across all executors now? 

This behavior is unique to the arrival rate executors, which are the only ones that have an internal VU pool. They introduce an additional meaning to "active" VUs, which can actually remain dormant until they're "running" an iteration. This change simply synchronizes the "running" counter to the global "active" one, which makes the value reported globally more accurate.

**Review by codebien — DISMISSED** [review]
LGTM :bow:

File: lib/executor/constant_arrival_rate_test.go [hunk 3]
```diff
@@ -354,3 +354,41 @@ func TestConstantArrivalRateGlobalIters(t *testing.T) {
...
+
+	var (
+		running          int64
+		getCurrActiveVUs func() int64
+		runMx            sync.Mutex
+	)
+
+	runner := simpleRunner(func(ctx context.Context, _ *lib.State) error {
+		runMx.Lock()
+		assert.Equal(t, atomic.AddInt64(&running, 1), getCurrActiveVUs())
```
Comment (mstoykov) [inline lib/executor/constant_arrival_rate_test.go:378]: nit: either the atomic here is pointless (we already have mutex) or you should also use it t o read the value in the checks at the end of the test

**Review by mstoykov — APPROVED** [review]
LGTM! That seems way too easy :rofl: but I expect it was hard to make it this easy :clap: 


I did some benchmarking and it seems to have some performance impact in one of my scripts lying around, but it seemed really small. After some more tests - specifically with empty default functions, I am pretty sure that was due to some randomness in my original script, so it seems to work really well.

File: lib/executor/constant_arrival_rate_test.go (hunk 3 above)
Comment (imiric) [reply lib/executor/constant_arrival_rate_test.go]: Good point. I removed the atomic in 7162f778.

The reason I didn't use it at the end is because the test is already done at that point, so it should be safe to read it directly.

**Comment by imiric** [comment]
> but I expect it was hard to make it this easy

Not really, I was surprised as well, and thought there might be a catch I'm missing. :sweat_smile: It does have the side effect that the global VU counter is now "bursty" as the arrival rate VU counter is, so we might want to cache it slightly at least in the UI, since currently it's not very legible on terminals with high refresh rate.

Thanks for confirming it doesn't impact performance. I was slightly concerned about that as well. :+1:

**Review by mstoykov — APPROVED** [review]

**Review by codebien — APPROVED** [review]
@imiric FYI, I triggered the stuck CLA check. Remember to squash and merge so we can aggregate the fixups. Thanks for doing this :bow:

**Comment by agilob** [comment]
Awesome 
image
